### PR TITLE
chore(las): Improving typings

### PR DESCRIPTION
**Pull request by morphoix**

### CODE REVIEW

File: modules/las/src/lib/laslaz-decoder.ts [hunk 1]
```diff
@@ -78,37 +96,66 @@ function parseLASHeader(arraybuffer) {
...
+  /**
+   * Reading data
+   * @param count
+   * @param skip
+   * @returns new ArrayBuffer, count, hasMoreData
+   */
+  readData(count: number, skip: number) {
     const {header, arraybuffer} = this;
-    if (!header) {
+    if (!this.header) {
```
Comment (dryabinin-actionengine) [inline modules/las/src/lib/laslaz-decoder.ts:153]: Why do you need `this.header` here? It is already taken from context above.

File: modules/las/src/lib/parse-las.ts [hunk 2]
```diff
@@ -151,8 +167,18 @@ export function parseLASChunked(rawData, skip, onParseData) {
...
+ * @param batchSize
+ * @param colorDepth
+ * @returns boolean
+ */
+function detectTwoByteColors(
+  decoder: any = {},
+  batchSize: number,
+  colorDepth?: number | string
+): boolean {
+  let twoByteColor: any | boolean;
```
Comment (dryabinin-actionengine) [inline modules/las/src/lib/parse-las.ts:181]: What's the point of using `any | bool` here? Maybe do `twoByteColor` `false` or `true` by default and make it bool type only?

File: modules/las/src/las-loader.ts [hunk 3]
```diff
@@ -6,13 +6,31 @@ import type {Loader, LoaderOptions} from '@loaders.gl/loader-utils';
...
 
+/**
+ * Type for header of the .las file
+ */
+export type LASHeader = {
+  pointsOffset: number;
+  pointsFormatId: number;
+  pointsStructSize: number;
+  pointsCount: number;
+  scale: number | any;
```
Comment (dryabinin-actionengine) [inline modules/las/src/las-loader.ts:24]: I don't see the value of `number | any` typing.

File: modules/las/src/lib/laslaz-decoder.ts (hunk 1 above)
Comment (morphoix) [reply modules/las/src/lib/laslaz-decoder.ts:153]: > Why do you need `this.header` here? It is already taken from context above.

fixed

File: modules/las/src/lib/parse-las.ts (hunk 2 above)
Comment (morphoix) [reply modules/las/src/lib/parse-las.ts:181]: > What's the point of using `any | bool` here? Maybe do `twoByteColor` `false` or `true` by default and make it bool type only?

ok

File: modules/las/src/las-loader.ts [hunk 4]
```diff
@@ -6,13 +6,31 @@ import type {Loader, LoaderOptions} from '@loaders.gl/loader-utils';
 const VERSION = typeof __VERSION__ !== 'undefined' ? __VERSION__ : 'latest';
 
 export type LASLoaderOptions = LoaderOptions & {
-  las?: {
+  las: {
```
Comment (ibgreen) [inline modules/las/src/las-loader.ts:9]: `options.las` should be optional. You should not need to provide options. Look at other loaders for how this is handled by applying defaults.

File: modules/las/src/lib/laslaz-decoder.ts [hunk 5]
```diff
@@ -3,46 +3,60 @@
...
 import getModule from '../libs/laz-perf';
 
-let Module = null;
+let Module: any = null;
+
+type LASReaders = {
+  [key: number]: (dv: DataView) => {
+    [LASAttribute: string]: number | number[];
+  };
+};
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:16]:
```suggestion
type LASReader = (dv: DataView) => {
    [LASAttribute: string]: number | number[];
type LASReaders = {
  [key: number]: LASReader;
  };
};
```

File: modules/las/src/lib/laslaz-decoder.ts [hunk 6]
```diff
@@ -78,37 +96,66 @@ function parseLASHeader(arraybuffer) {
...
+      mins: [0],
+      totalToRead: 0,
+      totalRead: 0,
+      versionAsString: '',
+      isCompressed: true
+    };
   }
-
+  /**
+   * Nothing needs to be done to open this
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:129]: I think this comment should remain inside the function. This comment is not a description of the semantics of this function, but rather an implementation comment.

File: modules/las/src/lib/laslaz-decoder.ts [hunk 7]
```diff
@@ -147,27 +194,41 @@ class LASLoader {
...
-//
+/**
+ * LAZ Loader
+ * Uses NaCL module to load LAZ files
+ */
 class LAZLoader {
-  constructor(arraybuffer) {
+  arraybuffer: ArrayBuffer;
+  instance: any;
+  header: LASHeader | null;
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:215]:
```suggestion
  instance: any = null;
  header: LASHeader | null = null;
```

File: modules/las/src/lib/laslaz-decoder.ts [hunk 8]
```diff
@@ -147,27 +194,41 @@ class LASLoader {
...
+
+  constructor(arraybuffer: ArrayBuffer) {
     this.arraybuffer = arraybuffer;
     this.instance = null; // LASZip instance
+    this.header = null;
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:220]:
```suggestion
```

File: modules/las/src/lib/laslaz-decoder.ts [hunk 9]
```diff
@@ -147,27 +194,41 @@ class LASLoader {
...
 
     if (!Module) {
       // Avoid executing laz-perf on import
       Module = getModule();
     }
   }
 
+  /**
+   * Opens the file using
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:229]: Incomplete comment.
```suggestion
   * Opens the file
```

File: modules/las/src/lib/laslaz-decoder.ts [hunk 10]
```diff
@@ -245,6 +306,10 @@ class LAZLoader {
     }
   }
 
+  /**
+   * Deletes the instance if it equals to null
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:310]: Confusing/incorrect comment
```suggestion
   * Deletes the instance
```

File: modules/las/src/lib/laslaz-decoder.ts [hunk 11]
```diff
@@ -272,7 +347,12 @@ class LASDecoder {
     this.maxs = header.maxs;
   }
 
-  getPoint(index) {
+  /**
+   * Decodes data depends on this point size
+   * @param index
+   * @returns
+   */
+  getPoint(index: number) {
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:355]: Add return types?

File: modules/las/src/lib/laslaz-decoder.ts [hunk 12]
```diff
@@ -282,10 +362,25 @@ class LASDecoder {
...
+  formatId: number;
+  loader: LASLoader | LAZLoader;
+  isCompressed: boolean;
+  isOpen: boolean;
+  version: number;
+  versionAsString: string;
+
+  constructor(arraybuffer: ArrayBuffer) {
     this.arraybuffer = arraybuffer;
+    this.formatId = 0;
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:379]: Nit: Initialize constant values in the declarations above.

File: modules/las/src/lib/laslaz-decoder.ts [hunk 13]
```diff
@@ -314,33 +412,55 @@ export class LASFile {
...
   }
 
-  readData(count, start, skip) {
+  /**
+   * @param count
+   * @param start
+   * @param skip
+   * @returns Data
+   */
+  readData(count: number, start: number, skip: number) {
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:448]: Add return types...

File: modules/las/src/lib/laslaz-decoder.ts [hunk 14]
```diff
@@ -314,33 +412,55 @@ export class LASFile {
...
   close() {
     if (this.loader.close()) {
       this.isOpen = false;
     }
   }
 
+  /**
+   * @returns LASDecoder
+   */
   getUnpacker() {
```
Comment (ibgreen) [inline modules/las/src/lib/laslaz-decoder.ts:464]:
```suggestion
  /**
   */
  getUnpacker(): LASDecoder {
```

File: modules/las/src/lib/parse-las.ts [hunk 15]
```diff
@@ -62,7 +76,7 @@ export default function parseLAS(arrayBuffer: ArrayBuffer, options?: LASLoaderOp
       positions[pointIndex * 3 + 1] = position[1] * scaleY + offsetY;
       positions[pointIndex * 3 + 2] = position[2] * scaleZ + offsetZ;
 
-      if (color) {
+      if (color && colors !== null) {
```
Comment (ibgreen) [inline modules/las/src/lib/parse-las.ts:79]:
```suggestion
      if (color && colors) {
```

**Review by ibgreen — APPROVED** [review]
Note that you can just apply the suggestions directly in the github UI, no need to edit the code.

After that you can land and make another PR to address the remaining comments.

File: modules/las/src/lib/parse-las.ts (hunk 15 above)
Comment (morphoix) [reply modules/las/src/lib/parse-las.ts]: @ibgreen  Ok, fixed everything in https://github.com/visgl/loaders.gl/pull/1603/commits/26f67c6e2949cd89b8c73f9807d1503604fe29ca

File: modules/las/src/index.ts [hunk 16]
```diff
@@ -13,10 +13,9 @@ export {LASWorkerLoader};
...
   ...LASWorkerLoader,
-  parse: async (arrayBuffer: ArrayBuffer, options?: LASLoaderOptions) =>
+  parse: async (arrayBuffer: ArrayBuffer, options: LASLoaderOptions) =>
     parseLAS(arrayBuffer, options),
-  parseSync: (arrayBuffer: ArrayBuffer, options?: LASLoaderOptions) =>
-    parseLAS(arrayBuffer, options)
+  parseSync: (arrayBuffer: ArrayBuffer, options: LASLoaderOptions) => parseLAS(arrayBuffer, options)
 };
-
+// @ts-ignore
```
Comment (ibgreen) [inline modules/las/src/index.ts:20]: If you need to ignore this you have a problem in your loader definition. Some missing field presumably.

File: modules/las/src/las-loader.ts [hunk 17]
```diff
@@ -9,10 +9,28 @@ export type LASLoaderOptions = LoaderOptions & {
   las?: {
     fp64?: boolean;
     skip?: number;
-    colorDepth?: number;
+    colorDepth?: number | string;
   };
 };
 
+/**
```
Comment (ibgreen) [inline modules/las/src/las-loader.ts:16]: Putting it here makes files import each other circularly. We already do this for the options but let's not make it worse. Keep it in a `./types` field.

**Review by ibgreen — APPROVED** [review]
Two issues remain, but we can merge as is.

**Pull request merged**